### ISSUE TITLE
Password-protected portfolio page still shows the portfolio items

### ISSUE THREAD

**Issue by fashuvo** [Automattic/wp-calypso#60135]
### Quick summary

If a page has the portfolio template selected and the page is password-protected, it still shows the portfolio items. Only the content added through the page editor is password-protected.

### Steps to reproduce

1. Activate any of [these themes](https://wordpress.com/themes/filter/feature:portfolio/) on the site.
2. Create a new page or edit any existing page.
3. Select the `portfolio template` for a page.
4. Update the page and view the page.

### What you expected to happen

All the page content, including the portfolio items, should be hidden.

### What actually happened

Any content added through the page editor is hidden, but the portfolio items are still visible.

Here's a screenshot showing the page content is hidden, but the portfolio items being visible:

![Image 2022-01-17 at 10 04 16 PM](https://user-images.githubusercontent.com/6534920/149803242-e58543af-36c5-4b10-bc00-7e59d086807f.jpg)

### Context

The issue was discovered in #4678070-zen, and I was able to reproduce it on both simple and atomic sites.

### Simple, Atomic or both?

Simple, Atomic

### Theme-specific issue?

https://wordpress.com/themes/filter/feature:portfolio/

### Reproducibility

Consistent

### Severity

One

### Available workarounds?

Yes, difficult to implement

### Workaround details

Here are the steps to the workaround I came up with:

1. Select the `default template` for the page.

2. Then add the `Blog Posts block` to the page.

3. Once you add the block, choose to show only `portfolio projects` from the right sidebar.

**Comment by zaguiini** [comment]
## Root cause

When rendering the portfolio template, the portfolio page itself is protected, but the portfolio items are not. As password protection applies only to the page level, the portfolio items, which are not password-secured, are displayed.

## Possible solutions

### Filter in Jetpack code

The idea was to add a `pre_get_posts` filter and check for password protection. If the post type in the query is `jetpack-portfolio` and the main post is password-protected, we would modify the query to ensure it doesn’t return any data, e.g.:

```php
if ( 'jetpack-portfolio' === $root_query->post_type ) {
    $query->set(‘post__in’, array(0));
}
```

We managed to hide the items that way, but it still displayed a “no results found” message. Ideally, it should show nothing.

It behaves that way because our hook lets the query normally run, and it only causes the returned result set to be empty. This solution has another disadvantage - it would hide password protection in code responsible for portfolio features, which would limit the customizations theme authors can currently make.

### Fix in themes

@wojtekn reviewed all ten themes available on [WordPress.com](http://wordpress.com/) marked as themes that include the “portfolio” feature.

There is one theme that does it correctly. Other themes could be improved by adding the missing check for password protection.

#### Blask theme
I couldn’t reproduce the issue in this theme. This theme uses the `post_password_required` function to check for the password; if required (or missing in a cookie), it displays the password form. Otherwise, it shows a portfolio. This theme works correctly.

#### Rebalance theme
I reproduced the issue in this theme - portfolio items are displayed despite having password protection enabled for the page.
However, this theme contains another issue. As it doesn’t display post content, the password protection form is not shown either. We could resolve the issue in a similar way to the Blask theme.

#### Snaps theme
This theme doesn’t support the Jetpack portfolio feature, although, in the [WordPress.com](http://wordpress.com/) directory, it is marked as supporting that. It would be good to consider removing the `feature: portfolio` flag from this theme.

#### Other themes

I reproduced the issue using all other themes: Lodestar, Orvis, Argent, Sketch, Sidespied, Espied, and Illustratr.
The issue in all of them can be fixed by modifying the template responsible for displaying portfolio and wrapping code responsible for displaying portfolio items with:

```php
<?php if ( ! post_password_required() ) : ?>
<?php /* code that displays portfolio */ ?>
<?php endif; ?>
```

## Ask for Feedback

I would appreciate @Automattic/theam and @Automattic/jetpack to review the solution and would appreciate some feedback. Is there a way to fix to ensure password protection at the plugin level -- that is, before reaching the template? Or will we need to modify all themes to match `Blask` behavior?

**Comment by zaguiini** [comment]
## Root cause

When rendering the portfolio template, the portfolio page itself is protected, but the portfolio items, [usually fetched in a separate query](https://themes.trac.wordpress.org/browser/orvis/1.1.3/page-templates/portfolio-page.php#L56-L96), are not. As password protection applies only to the page level, the portfolio items, which are not password-secured, are displayed. There is no additional check for not showing the portfolio items.

## Possible solutions

### Filter in Jetpack code

The idea was to add a `pre_get_posts` filter and check for password protection. If the post type in the query is `jetpack-portfolio` and the main post is password-protected, we would modify the query to ensure it doesn’t return any data, e.g.:

```php
if ( 'jetpack-portfolio' === $root_query->post_type ) {
    $query->set(‘post__in’, array(0));
}
```

We managed to hide the items that way, but it still displayed a “no results found” message. Ideally, it should show nothing.

It behaves that way because our hook lets the query normally run, and it only causes the returned result set to be empty. This solution has another disadvantage - it would hide password protection in code responsible for portfolio features, which would limit the customizations theme authors can currently make.

### Fix in themes

@wojtekn reviewed all ten themes available on [WordPress.com](http://wordpress.com/) marked as themes that include the “portfolio” feature.

There is one theme that does it correctly. Other themes could be improved by adding the missing check for password protection.

#### Blask theme
I couldn’t reproduce the issue in this theme. This theme uses the `post_password_required` function to check for the password; if required (or missing in a cookie), it displays the password form. Otherwise, it shows a portfolio. This theme works correctly.

#### Rebalance theme
I reproduced the issue in this theme - portfolio items are displayed despite having password protection enabled for the page.
However, this theme contains another issue. As it doesn’t display post content, the password protection form is not shown either. We could resolve the issue in a similar way to the Blask theme.

#### Snaps theme
This theme doesn’t support the Jetpack portfolio feature, although, in the [WordPress.com](http://wordpress.com/) directory, it is marked as supporting that. It would be good to consider removing the `feature: portfolio` flag from this theme.

#### Other themes

I reproduced the issue using all other themes: Lodestar, Orvis, Argent, Sketch, Sidespied, Espied, and Illustratr.
The issue in all of them can be fixed by modifying the template responsible for displaying portfolio and wrapping code responsible for displaying portfolio items with:

```php
<?php if ( ! post_password_required() ) : ?>
<?php /* code that displays portfolio */ ?>
<?php endif; ?>
```

## Ask for Feedback

We would appreciate @Automattic/theam and @Automattic/jetpack to review the solution and would appreciate some feedback. Is there a way to fix to ensure password protection at the plugin level -- that is, before reaching the template? Or will we need to modify all themes to match `Blask` behavior?

**Comment by jeherve** [comment]
I would personally vote for fixing this in the themes, it seems more logical since the template defined in the theme is already what's used to control what is displayed on those pages.

**Comment by zaguiini** [comment]
Creating a checklist so we can keep track of the progress in the themes...

### Password-protected portfolio pages

- [x] [Rebalance](https://github.com/Automattic/themes/pull/6008)
- [x] [Lodestar](https://github.com/Automattic/themes/pull/6011)
- [x] Orvis: D81025-code
- [x] Argent: D81081-code
- [x] [Sketch](https://github.com/Automattic/themes/pull/6012)
- [x] Sidespied: D81079-code
- [x] Espied: D81073-code
- [x] [Illustratr](https://github.com/Automattic/themes/pull/6013)
- [x] Snaps: D81082-code

These themes are all updated. We did some searching on the themes and that's the most comprehensive list we have at the moment. They're all fixed, so we're marking this issue as done and closing it.

### WPORG syncing (need further action)

None of the themes have been updated in WPORG for a while, and the attempts to update them after deploying a new version to WPCOM failed. They need to be updated and uploaded after the warnings and errors are fixed.

The Espied theme had an older version (1.2.2) running on WPCOM while WPORG has the latest version numbered 1.2.3. This caused a version collision as we just published 1.2.3 on WPCOM. To fix this, we need to sync both versions and bump them.
